index-scheduler/src/lru.rs

```diff
@@ -0,0 +1,202 @@
+    ///
+    /// The capacity is allocated up-front, and will never change through a [`Self::put`] operation.
+    ///
+    /// # Panics
+    ///
+    /// - If the capacity is 0.
+    /// - If the capacity exceeds `isize::MAX` bytes.
+    pub fn new(cap: usize) -> Self {
+        assert_ne!(cap, 0, "The capacity of a cache cannot be 0");
+        Self {
+            // Note: since the element of the vector contains an AtomicU64, it is definitely not zero-sized so cap will never be usize::MAX.
+            data: Vec::with_capacity(cap),
+            generation: AtomicU64::new(0),
+            cap,
+        }
+    }
+
+    /// The capacity of this LRU cache, that is the maximum number of elements it can hold before evicting elements from the cache.
+    ///
+    /// The cache will contain at most this number of elements at any given time.
+    pub fn capacity(&self) -> usize {
+        self.cap
+    }
+
+    fn next_generation(&self) -> u64 {
+        // Acquire so this "happens-before" any potential store to a data cell (with Release ordering)
+        let generation = self.generation.fetch_add(1, Ordering::Acquire);
+        generation + 1
+    }
+
+    fn next_generation_mut(&mut self) -> u64 {
+        let generation = self.generation.get_mut();
+        *generation += 1;
+        *generation
+    }
+
+    /// Add a value in the cache, evicting an older value if necessary.
+    ///
+    /// If a value was evicted from the cache, it is returned.
+    ///
+    /// # Complexity
+    ///
+    /// - If the cache is full, then linear in the capacity.
+    /// - Otherwise constant.
+    pub fn put(&mut self, value: T) -> Option<T> {
+        // no need for a memory fence: we assume that whichever mechanism provides us synchronization
+        // (very probably, a RwLock) takes care of fencing for us.
+
+        let next_generation = self.next_generation_mut();
+        let evicted = if self.is_full() { self.pop() } else { None };
+        self.data.push((AtomicU64::new(next_generation), value));
+        evicted
+    }
+
+    /// Evict the oldest value from the cache.
+    ///
+    /// If the cache is empty, `None` will be returned.
+    ///
+    /// # Complexity
+    ///
+    /// - Linear in the capacity of the cache.
+    pub fn pop(&mut self) -> Option<T> {
+        // Iterator::min_by_key provides shared references to its elements,
+        // but we need (and can afford!) an exclusive one, so let's make an explicit loop
+        let mut min_generation_index = None;
+        for (index, (generation, _)) in self.data.iter_mut().enumerate() {
+            let generation = *generation.get_mut();
+            if let Some((_, min_generation)) = min_generation_index {
+                if min_generation > generation {
+                    min_generation_index = Some((index, generation));
+                }
+            } else {
+                min_generation_index = Some((index, generation))
+            }
```

Comment:
Can't we replace that with a match, that's shorter? But I don't understand why you can't use the `min_by_key` here as it is not as issue right? As it returns a `Self::Item` in the end even if it was `&Self::Item` in the closure.

```suggestion
            match min_generation_index {
                Some((_, min_generation)) => if min_generation > *generation.get_mut() {
                    min_generation_index = Some((index, generation));
                },
                None => min_generation_index = Some((index, generation)),
            }
```

Reply:
Apologies, I see my doc comment was insufficient, and I had to think again what the full reason was.

The reason is performance.

min_by_key gives us a shared reference to the generation, which is an atomic integer. Atomic integers require an atomic load with an implicit memory fence to be correct. It is unnecessary because we have an exclusive reference, meaning either we have a single thread or we have some kind of lock that handled the memory fencing for us.

This is why atomic integers allow direct access "as if they were simple integers" when we can provide an exclusive reference to them. Which is what I use here.

Alternatively, I could have:

- performed loads with an acquire semantics, but that would be subefficient performance wise
- performed loads with a relaxed semantics, and put a an explicit memory fence above the loop, but even this fence is unnecessary
- performed loads with a relaxed semantics, and put a comment explaining why this is not a problem.

I preferred shunting completely the atomic load by using the `get_mut()` method that returns a exclusive reference to the atomic.

To be clear, here the important semantic distinction is exclusivity, not mutability of the `&mut`.

Feel free to tell me if:

1. the explanation is clearer now
2. if you prefer that I change for one of alternatives outlined above
3. how you'd like me to improve the doc comment so the reason becomes clearer. It could also be a module wide comment because I'm using the exclusiveness of `&mut AtomicU64` in quite a few locations.

Reply:
Changed the comment to:

```rust
        // Don't use `Iterator::min_by_key` that provides shared references to its elements,
        // so that we can get an exclusive one.
        // This allows to handles the `AtomicU64`s as normal integers without using atomic instructions.
```

Tell me what you think 👍

Reply:
Thank you very much, much clearer!